### DISCUSSION TITLE
Conditional on survey-text-response

### DISCUSSION THREAD

**Discussion by mshastry2022** [jspsych/jsPsych#846] [Q&A] [answered]
Hello everyone!

I am really new to JsPsych, and would love any help. I am trying to create an experiment that has someone enter a value through survey-text-response. They will be entering a numerical value from 0-100. They have 10 trials, so they will enter a new number 10 times. I am trying to find a way that provides feedback per trial for the number entered. For example, if someone were to enter in 50, I want them to see "-$." If someone were to enter 30, they will see "+$". One number, 60, is the winning number, where they will see "$$$". Some numbers have no feedback. 

I have an array of numbers that will lead to -$, another array of numbers for +$, and an third array of numbers that will lead to no feedback. How do I write the code that will provide such feedback? 

I have a bunch of code that sets up the survey, and lists out the arrays. I know my code does not connect with each part, so I am asking how I can provide feedback based on the survey answer, for each of the 10 trials. I know if statements are included, I am just not sure where. 

```js
var categorization = {
    type: 'survey-text',
    questions: [
      {prompt: 'Enter a number:', columns: 3, name: 'Number'},
  ],
}

conditional_function: function() {
  var data = jsPsych.data.getLastTrialData().values[0];
var positiveNumber = [70, 84, 65, 10, 23, 56, 52, 57, 69, 40, 6, 93, 25, 21,
        50, 49, 75, 22, 32, 13, 88, 80, 89, 42, 82, 79, 53, 92, 19, 48, 12, 99, 2];
  var negativeNumber = [15, 81, 31, 58, 8, 43, 20, 71, 11, 47, 73, 90, 62, 83,
        9, 39, 17, 18, 45, 78, 64, 63, 35, 87, 24, 100, 91, 7, 27, 55, 51, 94, 72];
  var neutralNumber = [5, 86, 26, 33, 95, 1, 98, 54, 14, 0, 44, 29, 74,
        97, 36, 28, 67, 59, 34, 85, 96, 77, 38, 30, 68, 37, 61, 4, 3, 76, 46, 66, 41,
        16];
  if(JSON.parse(data.responses).Q0 == 60) {
    return true;
  } else {
    return false;
  }
  }
  timeline.push(conditional_function);

  function myFunction() {
var trialNumber = 10;
var i;
for (i = 0, i < trialNumber, i++) {
  var number = prompt("Please enter a number", " ");
  var positiveNumber = [70, 84, 65, 10, 23, 56, 52, 57, 69, 40, 6, 93, 25, 21,
        50, 49, 75, 22, 32, 13, 88, 80, 89, 42, 82, 79, 53, 92, 19, 48, 12, 99, 2];
  var negativeNumber = [15, 81, 31, 58, 8, 43, 20, 71, 11, 47, 73, 90, 62, 83,
        9, 39, 17, 18, 45, 78, 64, 63, 35, 87, 24, 100, 91, 7, 27, 55, 51, 94, 72];
  var neutralNumber = [5, 86, 26, 33, 95, 1, 98, 54, 14, 0, 44, 29, 74,
        97, 36, 28, 67, 59, 34, 85, 96, 77, 38, 30, 68, 37, 61, 4, 3, 76, 46, 66, 41,
        16];
  if (positiveNumber.includes(number)) {
    document.getElementById("demo").innerHTML =
    "+$";
  }
  else if (negativeNumber.includes(number)) {
    document.getElementById("demo").innerHTML =
    "-$";
  }
  else if (neutralNumber.includes(number)) {
    document.getElementById("demo").innerHTML =
    "No gain or loss of monetary value";
  }
  else (number == 60) {
    document.getElementById("demo").innerHTML =
    "Congratulations! You found the winning number!";
  }
```

I will appreciate any kind of help!! 

Thank you in advance!!
Madhulika

**Comment by jodeleeuw** [comment, marked as answer]
Hi @mshastry2022,

I would recommend structuring it like this:

1. Use a conditional_function on a timeline that contains the feedback trial to determine if feedback should be shown at all.
2. If feedback should be shown, use [dynamic parameters](https://www.jspsych.org/overview/trial/#dynamic-parameters) to specify the feedback.

Here's a sketch of what the code could look like:

```js
var positiveNumber = [70, 84, 65, 10, 23, 56, 52, 57, 69, 40, 6, 93, 25, 21,
        50, 49, 75, 22, 32, 13, 88, 80, 89, 42, 82, 79, 53, 92, 19, 48, 12, 99, 2];
var negativeNumber = [15, 81, 31, 58, 8, 43, 20, 71, 11, 47, 73, 90, 62, 83,
        9, 39, 17, 18, 45, 78, 64, 63, 35, 87, 24, 100, 91, 7, 27, 55, 51, 94, 72];
var neutralNumber = [5, 86, 26, 33, 95, 1, 98, 54, 14, 0, 44, 29, 74, 97, 36, 
	28, 67, 59, 34, 85, 96, 77, 38, 30, 68, 37, 61, 4, 3, 76, 46, 66, 41, 16];

var categorization = {
    	type: 'survey-text',
    	questions: [
      		{prompt: 'Enter a number:', columns: 3, name: 'Number'},
 	 ],
	data: {task: 'guess-number'}
}

var feedback_trial = {
	type: 'html-keyboard-response',
	stimulus: function(){
		var guess = JSON.parse(jsPsych.data.get().filter({task: 'guess-number'}).last(1).values()[0].responses).Number
		if(positiveNumber.includes(guess)){
			return "<p>+$</p>";
		}
		if(negativeNumber.includes(guess)){
			return "<p>-$</p>";
		}
	},
	choices: jsPsych.NO_KEYS,
	trial_duration: 2000
}

var feedback = {
	timeline: [feedback_trial],
	conditional_function: function(){
		var guess = JSON.parse(jsPsych.data.get().filter({task: 'guess-number'}).last(1).values()[0].responses).Number
		if(neutralNumber.includes(guess)){
			return false;
		} else {
			return true;
		}
	}
}

var guessing_procedure = {
	timeline: [categorization, feedback],
	repetitions: 10
}

jsPsych.init({timeline: [guessing_procedure]});
```

One thing you may need to worry about here is what happens if someone types in non-numeric text. My current code won't handle that well, but you could check that in the `conditional_function`. You could also edit the `survey-text` plugin to use `<input type="number">` instead of text.

**Reply by mshastry2022** [reply]
Thank you so much! This was really helpful, and I am able to make my code work now!!

**Reply by legatrix** [reply]
Thanks for the tips. I was trying to get something similar, and I found when working with this example that I had to wrap the parsed JSON strings in `parseInt()` to make the code work.
